os/kernel/preference/preference_callback.c

```diff
@@ -0,0 +1,218 @@
+/****************************************************************************
+ *
+ * Copyright 2018 Samsung Electronics All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include <tinyara/config.h>
+#include <debug.h>
+#include <queue.h>
+#include <errno.h>
+#include <string.h>
+#include <mqueue.h>
+#include <signal.h>
+#include <fcntl.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <tinyara/preference.h>
+
+struct key_cb_list_s {
+	struct key_cb_list_s *flink;
+	char *key;
+	int type;
+	sq_queue_t cb_list; // node type : key_cb_node_t
+};
+typedef struct key_cb_list_s key_cb_list_t;
+
+struct key_cb_node_s {
+	struct key_cb_node_s *flink;
+	int pid;
+	preference_changed_cb cb_func;
+	void *cb_data;
+};
+typedef struct key_cb_node_s key_cb_node_t;
+
+static sq_queue_t g_key_cb_list; // node type : key_cb_list_t
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+static key_cb_list_t *preference_get_key_cb_list(int type, const char *key)
+{
+	key_cb_list_t *ptr;
+
+	ptr = (key_cb_list_t *)sq_peek(&g_key_cb_list);
+	while (ptr != NULL) {
+		if (ptr->type == type && !strncmp(ptr->key, key, strlen(ptr->key) + 1)) {
```

Comment:
No, some keys will be observed by number of applications. (# of app can check same key, for example low battery...)

Reply:
As you said, each task or thread should register callback for specific keys and then execute all callbacks of registered each tasks/pthreads when some value is changed from line 92.

Reply:
ah, I just confused, sorry for confusing you.